src/components/lab/Thermometer.js

```diff
@@ -0,0 +1,189 @@
+/* eslint-disable react/prop-types */
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { Shape, Group, Text, Rect } from 'react-konva';
+import {
+  THERMOMETER_RADIUS,
+  THERMOMETER_HEIGHT,
+  THERMOMETER_WIDTH,
+  THERMOMETER_POSITION_Y,
+  THERMOMETER_POSITION_X,
+  GRADUATION_HEIGHT,
+  GRADUATION_FONT_SIZE,
+  GRADUATION_PADDING_LEFT,
+  GRADUATION_WIDTH,
+  GRADUATION_LINE_HEIGHT,
+  THERMOMETER_COLOR,
+  THERMOMETER_STROKE_WIDTH,
+  THERMOMETER_STROKE_COLOR,
+  GRADUATION_MAX_NUMBER_TICKS,
+} from '../../config/constants';
+
+const ThermometerShape = ({
+  stroke,
+  strokeWidth,
+  fill,
+  height = THERMOMETER_HEIGHT,
+}) => {
+  // use trigonometry to get the exact angle matching
+  // the circle and the rectangle
+  const angle = Math.asin(THERMOMETER_WIDTH / 2 / THERMOMETER_RADIUS);
+
+  const drawThermometerShape = (context, shape) => {
+    const yOffset = THERMOMETER_POSITION_Y + THERMOMETER_HEIGHT - height;
+
+    context.beginPath();
+    context.lineTo(THERMOMETER_POSITION_X, yOffset);
+    context.lineTo(THERMOMETER_POSITION_X + THERMOMETER_WIDTH, yOffset);
+    context.lineTo(
+      THERMOMETER_POSITION_X + THERMOMETER_WIDTH,
+      yOffset + height,
+    );
+    context.arc(
+      THERMOMETER_POSITION_X + THERMOMETER_WIDTH / 2,
+      yOffset + height + THERMOMETER_RADIUS,
+      THERMOMETER_RADIUS,
+      -Math.PI / 2 + angle,
+      -angle + (3 / 2) * Math.PI,
+    );
+    context.lineTo(THERMOMETER_POSITION_X, yOffset + height);
+    context.closePath();
+    // (!) Konva specific method, it is very important
+    context.fillStrokeShape(shape);
+  };
+
+  return (
+    <>
+      <Shape
+        sceneFunc={(context, shape) => {
+          drawThermometerShape(context, shape, height);
+        }}
+        stroke={stroke}
+        strokeWidth={strokeWidth}
+        fill={fill}
+      />
+    </>
+  );
+};
+
+const Graduation = ({ from, to, currentTemperature }) => {
+  const x = THERMOMETER_POSITION_X + THERMOMETER_WIDTH;
+
+  // compute ideal step distance between ticks
+  const step = Math.ceil(
+    (Math.abs(to) + Math.abs(from)) / GRADUATION_MAX_NUMBER_TICKS,
+  );
+
+  // round min and max to closest graduation steps
+  const roundFrom = step * Math.floor(from / step);
+  const roundTo = step * Math.ceil(to / step) || step;
+
+  // compute graduation array at every step
+  const nbGraduation = (roundTo - roundFrom) / step + 1; // +1 to include max
+  let range = Array.from(
+    { length: nbGraduation },
+    (v, k) => k * step + roundFrom,
+  );
+
+  // select marks at most number of graduation we can display
+  const maxNbGraduation = Math.floor(THERMOMETER_HEIGHT / GRADUATION_HEIGHT);
+  const prop = Math.ceil(range.length / maxNbGraduation);
+  if (prop > 1) {
+    range = range.filter((_, i) => i % prop === 0);
+  }
+
+  // height between each tick
+  const deltaHeight = THERMOMETER_HEIGHT / range.length;
```

Comment:
this comment can be encapsulated within the variable name:

`const heightBetweenTicks = THERMOMETER_HEIGHT / range.length;`